docs/spec/distributing.rst

```diff
@@ -10,72 +10,154 @@ Stub files
 
 (Originally specified in :pep:`484`.)
 
-Stub files are files containing type hints that are only for use by
-the type checker, not at runtime.  There are several use cases for
-stub files:
+*Stub files*, also called *type stubs*, provide type information for untyped
+Python packages and modules. Stub files serve multiple purposes:
+
+* They are the only way to add type information to extension modules.
+* They can provide type information for packages that do not wish to
+  add them inline.
+* They can be distributed separately from the package or module that they
+  provide types for. The latter is referred to as the *implementation*.
+  This allows stubs to be developed at a different pace or by different
+  authors, which is especially useful when adding type annotations to
+  existing packages.
+* They can act as documentation, succinctly explaining the external
+  API of a package, without including implementation details or private
+  members.
+
+Stub files use a subset of the constructs used in Python source files, as
+described in :ref:`stub-file-supported-constructs` below. Type checkers should
+parse a stub that uses only such constructs without error and not interpret any
+construct in a manner contradictory to this specification. However, type
+checkers are not required to implement checks for all of these constructs and
+can elect to ignore unsupported ones. Additionally, type checkers can support
+constructs not described here.
+
+If a stub file is found for a module, the type checker should not read the
+corresponding "real" module. See :ref:`mro` for more information.
+
+.. _stub-file-syntax:
+
+Syntax
+^^^^^^
+
+Stub files are syntactically valid Python files in the earliest Python version
+that is not yet end-of-life. They use a ``.pyi`` suffix. The Python syntax used
+by stub files is independent from the Python versions supported by the
+implementation, and from the Python version the type checker runs under (if
+any).
+
+Type checkers should treat stubs as if ``from __future__ import annotations`` is
```

Comment:
This may be useful as a mnemonic but it's not a particularly precise description. For example, forward references in type aliases are valid in stubs, but they would fail at runtime even with the future import enabled. Moreover, `from __future__ import annotations` is about to be deprecated.

I would suggest something like:

> Type checkers should support all type system features in stub files, including those introduced in newer versions. For example, the pipe union syntax (``X | Y``) introduced in Python 3.10 may be used even before Python 3.9 reaches end-of-life. Type checkers should also support forward references in all contexts, without the need to quote strings. For example, forward references may be used in annotations and in the right-hand side of assignments used as type aliases.

Reply:
Hmm, I like the previous phrasing better. The proposed change doesn't fully capture the intended behavior. It's not just that type checkers should *support* forward references. They should *assume forward references within all type expressions* in a stub file, as they do when `from __future__ import annotations` is used in a source file. The distinction is important because symbol resolution rules change when assuming forward references.

Perhaps we should say that "all type expressions in a stub file should be evaluated as though they are quoted even if they are not".

Reply:
Let me think a bit about how to rephrase this. @JelleZijlstra I see what you mean about the current wording being incorrect, but supporting "all type system features" doesn't seem quite right either. For example, type checkers aren't expected to support the new `type` keyword (PEP 695) in stubs for a while yet, right, since it's new syntax in 3.12?

Reply:
A conformant type checker should definitely support the `type` keyword regardless of whether it's used in a ".pyi" or a ".py" file.

As a stub author, you can decide whether or not to use this feature based on whether the most popular type checkers support this feature, but this is not a concern of the typing spec.

Reply:
@erictraut What version of Python syntax do you think type checkers ought to support in stubs? Right now, this PR says that stubs use the syntax of the earliest Python version not yet end-of-life (currently 3.8). That's what was suggested [here](https://github.com/python/typing/issues/1605#issuecomment-1908644244), and I went with it because it seemed sensible and unambiguous. But if we stick with this, it would be contradictory to require type checkers to support newer syntax unless it's a case like the pipe union syntax, where `from __future__ import annotations` can be used to cheat, basically.

Reply:
The distinction I would make is between syntactic and non-syntactic features. For syntactic features (such as the `type` statement), stubs can only use features that are present in the oldest supported Python version. For non-syntactic type system features (such as the `|` operator, which does not change the parser), features from any Python version should be allowed.

Reply:
> stubs can only use features that are present in the oldest supported Python version

Why? I think this is good guidance for stub authors — and therefore should be added to the "guides". It's also understandable that typeshed contributors should be compelled to follow this rule. But it doesn't sound like something that belongs in the typing spec. The typing spec describes the type system and how type checkers should work. Conformant type checkers should support all of the latest syntax regardless of whether that's in stubs or source files.

Reply:
The reason to add this behavior to the spec is that it is part of the contract between stub authors and type checkers. A conformant type checker must accept `int | str` in a stub, even if it runs in Python 3.9 mode; a conformant type checker does not need to accept `type T = int` in a stub if it runs in Python 3.9 mode.

Reply:
> a conformant type checker does not need to accept type T = int in a stub if it runs in Python 3.9 mode.

I don't think that's documented anywhere in the typing spec currently. I don't remember seeing that, and I can't find any such statement in the spec.

This strikes me as an implementation limitation of mypy (because it chooses to use the parser built in to CPython). I think there are multiple compelling reasons why this implementation decision should be revisited by mypy maintainers, so I'd prefer not to codify this in the typing spec.

Reply:
The fact of the matter is that mypy – despite its shortcomings – is probably the most (explicitly) used and arguably most important type checker.[^1] In many cases, it's not super important (although of course not desirable) when the current behavior of mypy (or any other type checker for that matter) diverges from the spec, since users will usually use only one type checker for a particular project and can work around the specific idiosyncrasies. But in the case of library interfaces (which type stubs are), conformance is key. This is why I don't think the spec can just ignore mypy in this case – at least not as long there is no change in sight.[^2]

That said, I suggest that the `type` statement and potentially other newer syntax constructs should be allowed in type stubs, but support should be made optional. (A "MAY" in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119#section-5) parlance.) The guide can then recommend against using those constructs for the time being. The spec can periodically be updated to require support for certain syntax features as mypy (and other type checkers) catch up to support them unconditionally.

[^1]: pyright is such a fantastic type checker and I wish mypy would come close, especially when it comes to support for new features and response time for bug fixes.
[^2]: I also wish mypy would switch to an external ast parser and not rely on the ast module. Or at least significantly increase the minimum required Python version to run mypy.

Reply:
I took a stab at rewording this paragraph; PTAL.

I think it's important for the spec to explicitly state which version of Python syntax should be supported in stubs. Unlike for Python source files, the version isn't necessarily coupled to the runtime, so which one to use is potentially unclear. (Maybe this is already obvious to everyone, but it hadn't been stated yet, so I just wanted to make sure we're on the same page here.)

I agree with Sebastian that this version should be the earliest non-EOL'd version until the situation with mypy changes. One of the [purposes](https://typing.readthedocs.io/en/latest/spec/type-system.html#purpose) of the spec is to "provide library authors guarantees to rely on when working with multiple type checkers." The more the spec diverges from reality, the less useful it becomes for that purpose.

Reply:
I recognize and appreciate that mypy is important. For that reason, I think it's good for us to provide guidance to stub authors in the "guides" section that they should avoid the use of newer language features if they want to maximize compatibility.

I don't think it makes sense for the typing spec to say that conformant type checkers SHOULD support `type` statements in ".py" files, but MAY support `type` statements in stub files. Stub files shouldn't be special in this regard. Conformant type checkers should support all of the features in the latest published version of Python.

I'll note that mypy fails to conform with the typing spec [in a number of areas](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html). While that's not ideal, it's understandable. We should make sure that the community understands where mypy is not conformant so they can work around those limitations. Let's make sure we document this, but IMO this documentation doesn't belong in the typing spec.

One option is to amend the typing spec with some variant of what Jelle suggests above: "support for newer typing features that rely on grammar changes is optional if the type checker is implemented in Python and is running on an older version of Python." This statement would apply to both ".py" and ".pyi" files. I personally don't think the spec should say that because it would be codifying a current limitation in mypy. As a TC member, I would vote against such a change, but maybe other TC members would be in favor of it.

I understand why mypy originally leveraged the CPython parser, but that implementation decision has not aged well. And it's an implementation decision that could easily be revisited — especially now that the CPython code base uses the PEG parser. Mypy maintainers could incorporate the latest parser form the CPython code base if this were a priority.

Reply:
Hmm, I think we're at a bit of an impasse here. I understand where you're coming from, but it feels unnecessarily confusing to me for the spec and guides to say different things on supported syntax in stubs.

Discussion on this PR seems to have largely died down, so maybe it's a good time to ask the rest of the typing council to weigh in. I'll make a note of this and other unresolved points when opening the typing council issue.